server/src/main/java/org/elasticsearch/plugins/internal/documentreporting/DocumentReporter.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.plugins.internal.documentreporting;
+
+import org.elasticsearch.xcontent.XContentParser;
+
+/**
+ * An interface to allow wrapping an XContentParser and report that a document was parsed
+ * A default implementation returns a noop DocumentReporter - does not wrap a XContentParser and
+ * does not do anything upon reporting.
+ */
+public interface DocumentReporter {
+    /**
+     * a default noop implementation
+     */
+    DocumentReporter EMPTY_INSTANCE = new DocumentReporter() {
+    };
+
+    /**
+     * Decorates a provided xContentParser with additional logic (gather some state).
+     * The Decorator parser should use a state from DocumentReporter
+     * in order to allow a reportDocumentParsed method to use that state
+     * @param xContentParser to be decorated
+     * @return a decorator xContentParser
+     */
+    default XContentParser wrapParser(XContentParser xContentParser) {
+        return xContentParser;
+    }
+
+    /**
+     * Reports (perhaps to some external components) the state that was gathered by a decorated wrap
+     * @param indexName an index name to be reported along with the state
+     */
+    default void reportDocumentParsed(String indexName) {}
```

Comment:
I would have the index name at the start of the parse, rather than the end - so something like

    void wrapParser(XContentParser parser, String targetIndex);
    void finishParse();

Reply:
the point in https://github.com/elastic/elasticsearch/pull/97961#issuecomment-1653483209 I was trying to make was that we need to have an index parameter in DocumentReporter. It does not matter if it is in wrapParser or reportDocumentPrased (I don't mind adding it to wrapParser method)
if we were to rename this to `XContentParserWrapper` we would have to document that it needs an index name. For a generic name like this. How would you describe this?
I still think it is better to keep the intentions clear, that this is for reporting (without mentioning serverless in codebase0